File: apps/api/src/agent_eval_api/routers/v1/projects.py
"""Project endpoints — invoke Application use cases only."""

from __future__ import annotations

from agent_eval_application.commands.project import (
    CreateProjectCommand,
    DeprecateProjectCommand,
    RenameProjectCommand,
    UpdateProjectSettingsCommand,
)
from agent_eval_application.queries.queries import GetProjectQuery, ListProjectsQuery
from fastapi import APIRouter, Header, status

from agent_eval_api.dependencies import ActorDep, ContainerDep, ServicesDep
from agent_eval_api.schemas.common import CollectionResponse
from agent_eval_api.schemas.project import (
    CreateProjectRequest,
    ProjectResponse,
    RenameProjectRequest,
    UpdateProjectSettingsRequest,
)

router = APIRouter(prefix="/v1/projects", tags=["projects"])


@router.post(
    "",
    response_model=ProjectResponse,
    status_code=status.HTTP_201_CREATED,
    summary="Create Project",
    description="Create a Project. Optional Idempotency-Key enables safe retries.",
)
def create_project(
    body: CreateProjectRequest,
    actor: ActorDep,
    services: ServicesDep,
    container: ContainerDep,
    idempotency_key: str | None = Header(default=None, alias="Idempotency-Key"),
) -> ProjectResponse:
    dto = services.create_project.execute(
        CreateProjectCommand(
            actor=actor,
            name=body.name,
            description=body.description,
            settings=body.settings,
            idempotency_key=idempotency_key,
        )
    )
    # Grant Owner outside Application (membership is not a Domain concept).
    grant = getattr(container.auth, "grant", None)
    if callable(grant):
        grant(actor_id=actor.id, project_id=dto.id)
    return ProjectResponse.from_dto(dto)


@router.get(
    "",
    response_model=CollectionResponse[ProjectResponse],
    summary="List Projects",
    description="List Projects visible to the authenticated actor.",
)
def list_projects(
    actor: ActorDep,
    services: ServicesDep,
) -> CollectionResponse[ProjectResponse]:
    items = services.list_projects.execute(ListProjectsQuery(actor=actor))
    responses = [ProjectResponse.from_dto(p) for p in items]
    return CollectionResponse(items=responses, count=len(responses))


@router.get(
    "/{project_id}",
    response_model=ProjectResponse,
    summary="Get Project",
)
def get_project(
    project_id: str,
    actor: ActorDep,
    services: ServicesDep,
) -> ProjectResponse:
    dto = services.get_project.execute(
        GetProjectQuery(actor=actor, project_id=project_id)
    )
    return ProjectResponse.from_dto(dto)


@router.patch("/{project_id}", response_model=ProjectResponse)
def rename_project(
    project_id: str,
    body: RenameProjectRequest,
    actor: ActorDep,
    services: ServicesDep,
) -> ProjectResponse:
    dto = services.rename_project.execute(
        RenameProjectCommand(actor=actor, project_id=project_id, name=body.name)
    )
    return ProjectResponse.from_dto(dto)


@router.patch("/{project_id}/settings", response_model=ProjectResponse)
def update_project_settings(
    project_id: str,
    body: UpdateProjectSettingsRequest,
    actor: ActorDep,
    services: ServicesDep,
) -> ProjectResponse:
    dto = services.update_project_settings.execute(
        UpdateProjectSettingsCommand(
            actor=actor, project_id=project_id, settings=body.settings
        )
    )
    return ProjectResponse.from_dto(dto)


@router.post("/{project_id}/deprecate", response_model=ProjectResponse)
def deprecate_project(
    project_id: str,
    actor: ActorDep,
    services: ServicesDep,
) -> ProjectResponse:
    dto = services.deprecate_project.execute(
        DeprecateProjectCommand(actor=actor, project_id=project_id)
    )
    return ProjectResponse.from_dto(dto)

File: apps/web/src/lib/api/projects.ts
import { apiRequest } from "@/lib/api/client";

export type ProjectStatus = string;

export interface Project {
  id: string;
  name: string;
  description: string;
  status: ProjectStatus;
  created_at: string;
  settings: Record<string, string>;
}

export interface CollectionResponse<T> {
  items: T[];
  count: number;
  next_cursor: string | null;
  has_more: boolean;
}

export interface ListProjectsParams {
  cursor?: string;
  limit?: number;
  sort?: string;
  status?: string;
  q?: string;
}

export interface CreateProjectInput {
  name: string;
  description?: string;
  settings?: Record<string, string>;
}

function buildQuery(params: ListProjectsParams = {}): string {
  const search = new URLSearchParams();
  if (params.cursor) search.set("cursor", params.cursor);
  if (params.limit !== undefined) search.set("limit", String(params.limit));
  if (params.sort) search.set("sort", params.sort);
  if (params.status) search.set("status", params.status);
  if (params.q) search.set("q", params.q);
  const query = search.toString();
  return query ? `?${query}` : "";
}

export async function listProjects(
  token: string,
  params?: ListProjectsParams,
): Promise<CollectionResponse<Project>> {
  return apiRequest<CollectionResponse<Project>>(`/v1/projects${buildQuery(params)}`, {
    method: "GET",
    token,
  });
}

export async function getProject(token: string, projectId: string): Promise<Project> {
  return apiRequest<Project>(`/v1/projects/${encodeURIComponent(projectId)}`, {
    method: "GET",
    token,
  });
}

export async function createProject(
  token: string,
  input: CreateProjectInput,
  idempotencyKey?: string,
): Promise<Project> {
  const headers: Record<string, string> = {};
  if (idempotencyKey) {
    headers["Idempotency-Key"] = idempotencyKey;
  }
  return apiRequest<Project>("/v1/projects", {
    method: "POST",
    token,
    headers,
    body: {
      name: input.name,
      description: input.description ?? "",
      settings: input.settings ?? {},
    },
  });
}

export async function renameProject(
  token: string,
  projectId: string,
  name: string,
): Promise<Project> {
  return apiRequest<Project>(`/v1/projects/${encodeURIComponent(projectId)}`, {
    method: "PATCH",
    token,
    body: { name },
  });
}

export async function updateProjectSettings(
  token: string,
  projectId: string,
  settings: Record<string, string>,
): Promise<Project> {
  return apiRequest<Project>(`/v1/projects/${encodeURIComponent(projectId)}/settings`, {
    method: "PATCH",
    token,
    body: { settings },
  });
}

export async function deprecateProject(token: string, projectId: string): Promise<Project> {
  return apiRequest<Project>(`/v1/projects/${encodeURIComponent(projectId)}/deprecate`, {
    method: "POST",
    token,
  });
}
